dawn/docs/parallel_model.rst

```diff
@@ -0,0 +1,251 @@
+intermediate representation. It's specification can be found `here`_.
+
+Dependencies
+~~~~~~~~~~~~
+
+Generally, stencil computation in dawn differentiates between horizontal
+and the vertical dimension. The assumption is that individual statements
+of a stencil are embarrassingly parallel in the horizontal. For that
+reason, there is no guarantee that any horizontal point's computation
+will be finished before any other in each user statement.
+
+Vertical Dependencies
+^^^^^^^^^^^^^^^^^^^^^
+
+Each vertical region of a stencil specifies a loop order of execution.
+Dawn will uphold the following rules:
+
+-  If the loop order is forward, any statement with a vertical
+   dependency on other statements executed on a vertical level with
+   index ``k`` will be executed after each statement that it depends on
+   that appears previously in the current vertical region or in any
+   previous vertical region was executed on every vertical level with
+   indices ``k-1`` and smaller.
+-  If the loop order is backward, any statement with a vertical
+   dependency on another statement executed on a vertical level with
+   index ``k`` will be executed after each statement that it depends on
+   that appears previously in the current vertical region or in any
+   previous vertical region was executed on every vertical level with
+   indices ``k+1`` and bigger.
+
+An example of this would be the following
+
+.. code:: c++
+
+   1 vertical_region(k_start, k_end)
+   2   one = 1
+   3 vertical_region(k_start, k_end)
+   4   two = one[k - 1]
+
+Here, for every level with index ``k``, line two will be executed before
+line 4.
+
+Horizontal Dependencies
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Irrespective of the loop order every statement with a horizontal
+dependency on other statements will be executed after all statements it
+depends on are executed on the full horizontal plane in the same
+vertical level.
+
+An example of this is the following
+
+.. code:: c++
+
+   1 vertical_region(k_start, k_end)
+   2   one = 1
+   3   two = one[i - 1]
+
+For every vertical level, we guarantee that line two will be executed on
+the full horizontal plane before line three will be executed.
+
+--------------
+
+   It is currently unclear how we should handle this in one corner case:
+   Non-temporary fields can cause problem in the following case: write
+   center, read offset, write center. This is currently translated to
+   one multistage with three stages. The problem is that on input /
+   output fields we duplicate the writes which can have 2 side-effects
+   if neighboring blocks are out of sync: Block 1 writes in its extended
+   domain (line 2), block 2 writes (line 2) (same value), then reads
+   offset (line 3), then writes in the center (line 4). Now Block 1
+   reads (line 3) and gets the wrong value. This means that if we want
+   to solve the problem we need a global sync in that case (meaning
+   breaking the kernel call).
```

Comment:
Yes, this looks like the only possible solution. It should mean, if I'm correct, that line 4 must go in a new multistage. So we should fix the current implementation.
If we decide to remove the loop order concept from the multistage, than the multistage would still be useful because of this.

Reply:
Actually, for the frontend (and the high-level representation) we can easily accept read-write-read, because the problem only happens with temporary fields. As we only describe how the program behaves, we can easily think that each non-temporary field is first copied into a temporary field, and in the end copied out of the temporary field. So we can assume that each field has redundancy because there is a simple starting point from which we can optimize the problem. Do you agree?

Reply:
I agree that the frontend shouldn't change, but I don't understand the rest. Could you provide an example, please?

Reply:
Let me summarize and see also if I understood here, since I was confused about a couple of statements in the comments. 
@Stagno I think splitting into different multistages will not solve the problem, since multistages are not synchronizing among blocks which is the essential problem here. If splitting is the solution, it should be split into different stencils. 
However I think @lukasm91 gives the best solution, that is field version this into temporaries. A correction on his comment so that make sense to me, "because the problem only happens with **non-**temporary fields". 
That means we should have a mode of the field versioning that checks for this patterns and upon detection put the field into a temporary. Temporaries will have private allocations of redundant halos so that the current organization of a single multistage and multiple stages will work with the parallel model. Finally the temporary needs to be copied back into the field. 
Therefore the action here would be we need to support this in the field versioning pass, but the example is totally valid and we dont need to do anything concerning the parallel model. Correct?